### ISSUE TITLE
walker: Decouple WalkFunc to reduce walker dependencies

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform-ls#881]
Currently the module walker/indexer has many dependencies:

https://github.com/hashicorp/terraform-ls/blob/40a6ae8d9286d4bd7b891be1ed5ea0902a2da502/internal/terraform/module/walker.go#L63-L74

This makes it harder to change anything that involves the walker and harder to test.

We can extract the logic which happens when a module is found as `WalkFunc` and have that function alone depend on Job Scheduler etc.

https://github.com/hashicorp/terraform-ls/blob/40a6ae8d9286d4bd7b891be1ed5ea0902a2da502/internal/terraform/module/walker.go#L199-L200

**Comment by radeksimko** [comment]
https://github.com/hashicorp/terraform-ls/compare/f-walker-refactoring represents a very rough idea of how this could be implemented, but it still needs some further thinking about watcher-related indexing logic, which is likely to change as part of https://github.com/hashicorp/terraform-ls/issues/867 - specifically in the sense that walker should no longer need to be explicitly adding each walked path to the watcher as LSP-based watching supports wildcards.

With that in mind I'm putting this on hold until https://github.com/hashicorp/terraform-ls/pull/790 is finally reviewed & merged and revisit with https://github.com/hashicorp/terraform-ls/issues/867 in mind.

**Comment by radeksimko** [comment]
Currently blocked by https://github.com/hashicorp/terraform-ls/pull/953

**Comment by radeksimko** [comment]
Also it would be best to first finish & merge https://github.com/hashicorp/terraform-ls/pull/924
